File: sdk/go/metrics.go
package marmot

import (
	"context"

	apiclient "github.com/marmotdata/marmot/sdk/go/internal/gen/client"
	"github.com/marmotdata/marmot/sdk/go/internal/gen/client/metrics"
	"github.com/marmotdata/marmot/sdk/go/internal/gen/models"
)

// TotalAssets is the catalog-wide asset count.
type TotalAssets = models.V1MetricsTotalAssetsResponse

// AssetsByType is the by-type asset breakdown.
type AssetsByType = models.V1MetricsAssetsByTypeResponse

// AssetsByProvider is the by-provider asset breakdown.
type AssetsByProvider = models.V1MetricsAssetsByProviderResponse

// AssetCount is one (asset, count) entry.
type AssetCount = models.MetricsAssetCount

// QueryCount is one (query, count) entry.
type QueryCount = models.MetricsQueryCount

// TopOptions sets the time window and limit for top-N queries.
// Start and End must be RFC3339 timestamps.
type TopOptions struct {
	Start string
	End   string
	Limit int64
}

// MetricsService returns catalog usage and asset breakdown metrics.
type MetricsService struct {
	gen *apiclient.Marmot
}

// TotalAssets returns the total number of assets in the catalog.
func (s *MetricsService) TotalAssets(ctx context.Context) (*TotalAssets, error) {
	p := metrics.NewGetMetricsAssetsTotalParams().WithContext(ctx)
	resp, err := s.gen.Metrics.GetMetricsAssetsTotal(p)
	if err != nil {
		return nil, mapErr(err)
	}
	return resp.Payload, nil
}

// AssetsByType returns asset counts grouped by type.
func (s *MetricsService) AssetsByType(ctx context.Context) (*AssetsByType, error) {
	p := metrics.NewGetMetricsAssetsByTypeParams().WithContext(ctx)
	resp, err := s.gen.Metrics.GetMetricsAssetsByType(p)
	if err != nil {
		return nil, mapErr(err)
	}
	return resp.Payload, nil
}

// AssetsByProvider returns asset counts grouped by provider.
func (s *MetricsService) AssetsByProvider(ctx context.Context) (*AssetsByProvider, error) {
	p := metrics.NewGetMetricsAssetsByProviderParams().WithContext(ctx)
	resp, err := s.gen.Metrics.GetMetricsAssetsByProvider(p)
	if err != nil {
		return nil, mapErr(err)
	}
	return resp.Payload, nil
}

// TopAssets returns the most-viewed assets in the given time range.
func (s *MetricsService) TopAssets(ctx context.Context, opts TopOptions) ([]*AssetCount, error) {
	p := metrics.NewGetMetricsTopAssetsParams().WithContext(ctx).WithStart(opts.Start).WithEnd(opts.End)
	if opts.Limit > 0 {
		p = p.WithLimit(&opts.Limit)
	}
	resp, err := s.gen.Metrics.GetMetricsTopAssets(p)
	if err != nil {
		return nil, mapErr(err)
	}
	return resp.Payload, nil
}

// TopQueries returns the most-run queries in the given time range.
func (s *MetricsService) TopQueries(ctx context.Context, opts TopOptions) ([]*QueryCount, error) {
	p := metrics.NewGetMetricsTopQueriesParams().WithContext(ctx).WithStart(opts.Start).WithEnd(opts.End)
	if opts.Limit > 0 {
		p = p.WithLimit(&opts.Limit)
	}
	resp, err := s.gen.Metrics.GetMetricsTopQueries(p)
	if err != nil {
		return nil, mapErr(err)
	}
	return resp.Payload, nil
}

File: sdk/python/src/marmot/resources/metrics.py
"""Catalog usage and asset breakdown metrics."""

from __future__ import annotations

from typing import cast

from marmot._adapter import unwrap
from marmot._gen.api.metrics import (
    get_metrics_assets_by_provider,
    get_metrics_assets_by_type,
    get_metrics_assets_total,
    get_metrics_top_assets,
    get_metrics_top_queries,
)
from marmot._gen.client import AuthenticatedClient
from marmot._gen.models.asset_count import AssetCount
from marmot._gen.models.assets_by_provider_response import AssetsByProviderResponse
from marmot._gen.models.assets_by_type_response import AssetsByTypeResponse
from marmot._gen.models.query_count import QueryCount
from marmot._gen.models.total_assets_response import TotalAssetsResponse
from marmot._gen.types import UNSET, Unset


class MetricsResource:
    def __init__(self, client: AuthenticatedClient) -> None:
        self._c = client

    def total_assets(self) -> TotalAssetsResponse:
        """Return the total number of assets in the catalog."""
        return cast(
            TotalAssetsResponse,
            unwrap(get_metrics_assets_total.sync_detailed(client=self._c)),
        )

    def assets_by_type(self) -> AssetsByTypeResponse:
        """Return asset counts grouped by type."""
        return cast(
            AssetsByTypeResponse,
            unwrap(get_metrics_assets_by_type.sync_detailed(client=self._c)),
        )

    def assets_by_provider(self) -> AssetsByProviderResponse:
        """Return asset counts grouped by provider."""
        return cast(
            AssetsByProviderResponse,
            unwrap(get_metrics_assets_by_provider.sync_detailed(client=self._c)),
        )

    def top_assets(
        self,
        *,
        start: str,
        end: str,
        limit: int | None = None,
    ) -> list[AssetCount]:
        """Return the most-viewed assets in [start, end] (RFC3339 timestamps)."""
        limit_arg: int | Unset = limit if limit is not None else UNSET
        return cast(
            list[AssetCount],
            unwrap(
                get_metrics_top_assets.sync_detailed(
                    client=self._c, start=start, end=end, limit=limit_arg
                )
            ),
        )

    def top_queries(
        self,
        *,
        start: str,
        end: str,
        limit: int | None = None,
    ) -> list[QueryCount]:
        """Return the most-run queries in [start, end] (RFC3339 timestamps)."""
        limit_arg: int | Unset = limit if limit is not None else UNSET
        return cast(
            list[QueryCount],
            unwrap(
                get_metrics_top_queries.sync_detailed(
                    client=self._c, start=start, end=end, limit=limit_arg
                )
            ),
        )
